src/extensions/agent_modules/shared/upd_utils/src/adu_upd_utils.c

```diff
@@ -0,0 +1,104 @@
+#include "aduc/adu_upd_utils.h"
+#include <aduc/adu_module_state.h> // ADUC_MQTT_CLIENT_MODULE_STATE
+#include <aduc/adu_mqtt_protocol.h> // ADU_RESPONSE_MESSAGE_RESULT_CODE_*
+#include <aduc/agent_state_store.h>
+#include <aduc/logging.h>
+#include <aduc/mqtt_broker_common.h> // ADUC_MQTT_Message_Context
+#include <aduc/string_c_utils.h> // ADUC_Safe_StrCopyN
+#include <aduc/workqueue.h> // WorkQueueHandle
+#include <string.h> // strlen
+
+// keep this last to avoid interfering with system headers
+#include <aduc/aduc_banned.h>
+
+/**
+ * @brief Gets the update data object from the agent info operation context.
+ * @param context The update operation context.
+ * @return ADUC_Update_Request_Operation_Data* The update data object.
+ */
+ADUC_Update_Request_Operation_Data* UpdateData_FromOperationContext(ADUC_Retriable_Operation_Context* context)
+{
+    if (context == NULL || context->data == NULL)
+    {
+        Log_Error("Null input (context:%p, data:%p)", context, context->data);
+        return NULL;
+    }
+
+    return (ADUC_Update_Request_Operation_Data*)(context->data);
+}
+
+/**
+ * @brief Gets the update workqueue handle from the callback user object.
+ * @param obj The callback's user object.
+ * @return ADUC_Retriable_Operation_Context* The retriable operation context.
+ */
+WorkQueueHandle WorkQueueHandleFromCallbackUserObj(void* obj)
+{
+    ADUC_MQTT_CLIENT_MODULE_STATE* ownerModuleState = (ADUC_MQTT_CLIENT_MODULE_STATE*)obj;
+    ADUC_WorkQueues* workQueues = (ADUC_WorkQueues*)(ownerModuleState->moduleInitData);
+    return workQueues->updateWorkQueue;
```

Comment:
Suggestion: NULL check and gracefully return null? 

Same in 2 functions below.

Reply:
fixed with commit: [handle nulls](https://github.com/Azure/iot-hub-device-update/pull/585/commits/aa0f584e401c72e2c5ed5e37ebbec479659e6ae0)